index.d.ts

```diff
@@ -1,49 +1,68 @@
-import * as fastify from "fastify";
-import { IncomingMessage, Server, ServerResponse } from "http";
-import { DecodeOptions, Secret, SignOptions, VerifyCallback, VerifyOptions } from "jsonwebtoken";
-
-declare module "fastify"
-{
-    interface Jwt
-    {
-        decode: (token: string, options?: DecodeOptions) => null | { [key: string]: any } | string;
-        options: {
-            /**
-             * decodeOptions
-             */
-            decode: DecodeOptions,
-
-            /**
-             * signOptions
-             */
-            sign: SignOptions,
-
-            /**
-             * verifyOptions
-             */
-            verify: VerifyOptions,
-        };
-        secret: Secret;
-        sign: (playload: string | Buffer | object, options?: SignOptions, callback?: VerifyCallback) => string;
-        verify: (token: string, options?: VerifyOptions, callback?: VerifyCallback) => Promise<object | string>;
-    }
-    interface FastifyInstance<HttpServer = Server, HttpRequest = IncomingMessage, HttpResponse = ServerResponse>
-    {
-        jwt: Jwt;
-    }
+import * as fastify from 'fastify';
+import * as http from 'http';
+import * as jwt from 'jsonwebtoken';
+
+declare module 'fastify' {
+  namespace JWTTypes {
+    type PayloadType = object | string;
+
+    interface SignCallback extends jwt.SignCallback {}
 
-    interface FastifyRequest<
-        HttpRequest,
-        Query = DefaultQuery,
-        Params = DefaultParams,
-        Headers = DefaultHeaders,
-        Body = DefaultBody
-        >
-    {
-        jwtVerify(callback?: VerifyCallback): Promise<any>;
+    interface VerifyCallback<Decoded extends PayloadType> extends jwt.VerifyCallback {
+      (err: jwt.VerifyErrors, decoded: Decoded): void;
     }
+  }
+
+  interface JWT {
+    options: {
+      decode: jwt.DecodeOptions;
+      sign: jwt.SignOptions;
+      verify: jwt.VerifyOptions;
+    };
+    secret: jwt.Secret;
+
+    sign(payload: JWTTypes.PayloadType | Buffer, options?: jwt.SignOptions): string;
+    sign(payload: JWTTypes.PayloadType | Buffer, callback: JWTTypes.SignCallback): void;
+    sign(payload: JWTTypes.PayloadType | Buffer, options: jwt.SignOptions, callback: JWTTypes.SignCallback): void;
+
+    verify<Decoded extends JWTTypes.PayloadType = any>(token: string, options?: jwt.VerifyOptions): Decoded;

```

Comment:
What is the purpose of defaulting this to `any` if you are constraining it to `JWTTypes.PayloadType`?

Reply:
`PayloadType` is `object | string`, but `object | string` type requires type assertion for use.
For example,

```ts
{
  const token = app.jwt.sign({ key: 'value' });
  const decoded = app.jwt.verify(token); // <= object | string
  console.log(decoded.key);
}
```

throw Error like below.

```
Property 'key' does not exist on type 'string | object'.
  Property 'key' does not exist on type 'string'.
```

So if default value type is `object | string`, you should set type assertion.
It is hard to use, I think.

For convenience, I set `any` type as default.
When you want to check decoded type strictly, you can use generics.

```ts
interface Data {
  key: string;
}

{
  const orig: Data = { key: 'value' };
  const token = app.jwt.sign(orig);
  const decoded = app.jwt.verify<Data>(token);
  console.log(decoded.key);
}
```

Reply:
I have to disagree about this. You use TypeScript to strictly type your JavaScript code. By allowing things to be `any` like this you're opening yourself back up to the exact problems TypeScript is used to mitigate. If a user wishes to inline their `sign` payload then they can extend the JWT type interface and overload the method `payload` parameter. Furthermore, the `jsonwebtoken` types are strict about method payload types (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9fc7f6fbbdb741ea15c46b0778dd3595d483da26/types/jsonwebtoken/index.d.ts#L133)

Reply:
I'm not good at English, so I cannot completely understand what you want to say.
But I agree that we should avoid using `any` type.

Which should I do?

1. Revert types (Not use Generics)
2. Remain Generics but remove any type (like below)
    ```ts
    verify<Decoded extends JWTTypes.PayloadType = JWTTypes.PayloadType>(token: string, options?: jwt.VerifyOptions): Decoded;
    ```

Reply:
You should keep generics without the assignment
```typescript
verify<Decoded extends JWTTypes.PayloadType>(token: string, options?: jwt.VerifyOptions): Decoded;
```

Reply:
Ok, I'll fix it.